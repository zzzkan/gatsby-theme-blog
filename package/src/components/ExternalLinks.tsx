import React from "react"
import { HStack, Icon, Link, Tooltip } from "@chakra-ui/react"
import { IconType } from "@react-icons/all-files/lib"
import { FaLink } from "@react-icons/all-files/fa/FaLink"
import { FaRss } from "@react-icons/all-files/fa/FaRss"
import { FaGithub } from "@react-icons/all-files/fa/FaGithub"
import { FaTwitter } from "@react-icons/all-files/fa/FaTwitter"

const icon = (name: string): IconType | undefined => {
  if (name === "GitHub") return FaGithub
  if (name === "Twitter") return FaTwitter
  if (name === "RSS") return FaRss
  return FaLink
}

export type ExternalLinksProps = {
  readonly links: Array<{
    readonly name: string
    readonly url: string
    readonly label?: string
  }>
}

export const ExternalLinks: React.FC<ExternalLinksProps> = ({ links }) => {
  if (links == null || links.length === 0) return null
  return (
    <HStack spacing={2}>
      {links.map((link, index) => (
        <Tooltip
          key={`${link.name}-${index}`}
          label={link.label ?? link.name}
          placement={"top"}
          color={"inherit"}
          background={"secondaryBackground"}
        >
          <Link
            href={link.url}
            aria-label={`Transition to ${link.label ?? link.name}`}
            _hover={{ textDecoration: "none" }}
          >
            <Icon as={icon(link.name)} fontSize={"2xl"} />
          </Link>
        </Tooltip>
      ))}
    </HStack>
  )
}
