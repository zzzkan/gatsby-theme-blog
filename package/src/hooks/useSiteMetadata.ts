import { useStaticQuery, graphql } from "gatsby"

type SiteMetadataProps = {
  readonly title: string
  readonly siteUrl: string
  readonly description: string
  readonly author: string
  readonly publicationYear: string
  readonly image: string
}

export const useSiteMetadata = (): SiteMetadataProps => {
  const { site } = useStaticQuery<{
    site: { siteMetadata: SiteMetadataProps }
  }>(
    graphql`
      query {
        site {
          siteMetadata {
            title
            siteUrl
            description
            author
            publicationYear
            image
          }
        }
      }
    `
  )
  return site.siteMetadata
}
