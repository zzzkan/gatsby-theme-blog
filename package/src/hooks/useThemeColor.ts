import { useColorModeValue } from "@chakra-ui/react"

type ThemeColorProps = {
  tint: string
  primaryText: string
  secondaryText: string
  onTintText: string
  primaryBackground: string
  secondaryBackground: string
  codeBackground: string
  highlightBackground: string
}

export const useThemeColor = (): ThemeColorProps => {
  return {
    tint: useColorModeValue("brand.600", "brand.200"),
    primaryText: useColorModeValue("base.800", "base.100"),
    secondaryText: useColorModeValue("base.600", "base.400"),
    onTintText: useColorModeValue("white", "base.900"),
    primaryBackground: useColorModeValue("white", "base.900"),
    secondaryBackground: useColorModeValue(
      "RGBA(0, 0, 0, 0.02)",
      "RGBA(255, 255, 255, 0.02)"
    ),
    codeBackground: useColorModeValue(
      "RGBA(0, 0, 0, 0.02)",
      "RGBA(255, 255, 255, 0.02)"
    ),
    highlightBackground: useColorModeValue(
      "RGBA(0, 0, 0, 0.03)",
      "RGBA(255, 255, 255, 0.04)"
    ),
  }
}
