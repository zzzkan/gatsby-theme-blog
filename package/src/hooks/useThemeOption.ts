import { useStaticQuery, graphql } from "gatsby"
import { type ThemeOptionType } from "../types/themeOptionType"

export const useThemeOption = (): ThemeOptionType => {
  const data = useStaticQuery<{ themeOption: ThemeOptionType }>(graphql`
    query {
      themeOption(id: { eq: "@zzzkan/gatsby-theme-blog" }) {
        links {
          name
          url
          label
        }
      }
    }
  `)
  return data.themeOption
}
