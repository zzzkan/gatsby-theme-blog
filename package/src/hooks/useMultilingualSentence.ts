import { useThemeOption } from "./useThemeOption"
import { type multilingualSentenceType } from "../types/multilingualSentenceType"

export const useMultilingualSentence = (): multilingualSentenceType => {
  const { lang } = useThemeOption()
  return {
    getFeaturedImageAlt: () => {
      if (lang === "ja") return `アイキャッチ画像`
      return `Featured image`
    },
    getFeaturedImageLinkAlt: (str: string) => {
      if (lang === "ja") return `記事へ移動 (${str})`
      return `Go to the page (${str})`
    },
    getToggleColorModeLabel: () => {
      if (lang === "ja") return "カラーモードを変更"
      return "Switch to color mode"
    },
    getAllPostsText: () => {
      if (lang === "ja") return "記事一覧"
      return "All Posts"
    },
    getPreviousPageText: () => {
      if (lang === "ja") return "前の記事"
      return "Previous Post"
    },
    getNextPageText: () => {
      if (lang === "ja") return "次の記事"
      return "Next Post"
    },
    getRelatedPostsText: () => {
      if (lang === "ja") return "関連記事"
      return "Related Posts"
    },
    getPageCountText: (count: number) => {
      if (lang === "ja") return `${count}ページ目`
      return `page ${count}`
    },
    getTagPostsDescriptionText: (tag: string) => {
      if (lang === "ja") return `${tag}タグが付いた記事一覧`
      return `Posts tagged with ${tag}`
    },
  }
}
