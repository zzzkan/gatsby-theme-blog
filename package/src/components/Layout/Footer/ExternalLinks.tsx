import React from "react"
import { HStack, Icon, Link, Tooltip } from "@chakra-ui/react"
import { IconType } from "@react-icons/all-files/lib"
import { FaLink } from "@react-icons/all-files/fa/FaLink"
import { FaRss } from "@react-icons/all-files/fa/FaRss"
import { FaGithub } from "@react-icons/all-files/fa/FaGithub"
import { FaTwitter } from "@react-icons/all-files/fa/FaTwitter"
import { useThemeOption } from "../../../hooks/useThemeOption"
import { useThemeColor } from "../../../hooks/useThemeColor"

const icon = (name: string): IconType | undefined => {
  if (name === "GitHub") return FaGithub
  if (name === "Twitter") return FaTwitter
  if (name === "RSS") return FaRss
  if (name === "Profile") return FaLink
  return undefined
}

const ExternalLinks: React.FC = () => {
  const { links } = useThemeOption()
  const { secondaryBackground } = useThemeColor()
  return (
    <HStack spacing={2}>
      {links?.map((link) => (
        <Tooltip
          key={link.url}
          label={link.label ?? link.name}
          placement={"top"}
          color={"inherit"}
          background={secondaryBackground}
        >
          <Link
            href={link.url}
            aria-label={link.label ?? link.name}
            _hover={{ textDecoration: "none" }}
          >
            <Icon as={icon(link.name)} fontSize={"2xl"} />
          </Link>
        </Tooltip>
      ))}
    </HStack>
  )
}

export default ExternalLinks
