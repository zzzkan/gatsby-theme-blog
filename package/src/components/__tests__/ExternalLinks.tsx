import React from "react"
import { render } from "@testing-library/react"
import { ExternalLinks } from "../ExternalLinks"
import { FaLink } from "@react-icons/all-files/fa/FaLink"
import { FaRss } from "@react-icons/all-files/fa/FaRss"
import { FaGithub } from "@react-icons/all-files/fa/FaGithub"
import { FaTwitter } from "@react-icons/all-files/fa/FaTwitter"
import { useThemeOption } from "../../hooks/useThemeOption"

describe("ExternalLinks component", () => {
  vi.mock("@react-icons/all-files/fa/FaLink")
  vi.mock("@react-icons/all-files/fa/FaRss")
  vi.mock("@react-icons/all-files/fa/FaGithub")
  vi.mock("@react-icons/all-files/fa/FaTwitter")
  vi.mock("../../hooks/useThemeOption")

  const mockFaLink = FaLink as jest.Mock
  const mockFaRss = FaRss as jest.Mock
  const mockFaGithub = FaGithub as jest.Mock
  const mockFaTwitter = FaTwitter as jest.Mock
  const mockUseThemeOption = useThemeOption as jest.Mock

  afterEach(() => {
    vi.clearAllMocks()
  })

  afterAll(() => {
    vi.restoreAllMocks()
  })

  test("snapshot", () => {
    ;(useThemeOption as jest.Mock).mockImplementationOnce(() => {
      return {
        links: [
          {
            name: "Profile",
            url: "/profile",
            label: "Profile Label",
          },
          {
            name: "RSS",
            url: "/rss",
            label: "RSS Label",
          },
          {
            name: "GitHub",
            url: "/github",
            label: "GitHub Label",
          },
          {
            name: "Twitter",
            url: "/twitter",
            label: "Twitter Label",
          },
        ],
      }
    })
    const { asFragment } = render(<ExternalLinks />)
    expect(asFragment()).toMatchSnapshot()
  })

  test("should not render anything if links is null", () => {
    mockUseThemeOption.mockImplementationOnce(() => {
      return { links: null }
    })
    const { container } = render(<ExternalLinks />)
    expect(container.firstChild).toBeNull()
  })

  test("should not render anything if links is []", () => {
    mockUseThemeOption.mockImplementationOnce(() => {
      return { links: [] }
    })
    const { container } = render(<ExternalLinks />)
    expect(container.firstChild).toBeNull()
  })

  test("should render links.url as lnik", () => {
    mockUseThemeOption.mockImplementationOnce(() => {
      return {
        links: [
          {
            name: "",
            url: "/profile",
          },
        ],
      }
    })
    const { getByRole } = render(<ExternalLinks />)
    expect(getByRole("link")).toHaveAttribute("href", "/profile")
  })

  test("should render links.name as label", () => {
    mockUseThemeOption.mockImplementationOnce(() => {
      return {
        links: [
          {
            name: "Profile",
            url: "/",
          },
        ],
      }
    })
    const { getByLabelText } = render(<ExternalLinks />)
    expect(getByLabelText("Profile")).toBeInTheDocument()
  })

  test("should render links.label as label", () => {
    mockUseThemeOption.mockImplementationOnce(() => {
      return {
        links: [
          {
            name: "Profile",
            url: "/",
            label: "Profile Label",
          },
        ],
      }
    })
    const { queryByLabelText } = render(<ExternalLinks />)
    expect(queryByLabelText("Profile")).toBeNull()
    expect(queryByLabelText("Profile Label")).toBeInTheDocument()
  })

  test("should render FaLink", () => {
    ;(useThemeOption as jest.Mock).mockImplementationOnce(() => {
      return {
        links: [
          {
            name: "",
            url: "/",
          },
        ],
      }
    })
    render(<ExternalLinks />)
    expect(mockFaLink).toHaveBeenCalled()
    expect(mockFaRss).not.toHaveBeenCalled()
    expect(mockFaGithub).not.toHaveBeenCalled()
    expect(mockFaTwitter).not.toHaveBeenCalled()
  })

  test("should render FaRss", () => {
    ;(useThemeOption as jest.Mock).mockImplementationOnce(() => {
      return {
        links: [
          {
            name: "RSS",
            url: "/",
          },
        ],
      }
    })
    render(<ExternalLinks />)
    expect(mockFaLink).not.toHaveBeenCalled()
    expect(mockFaRss).toHaveBeenCalled()
    expect(mockFaGithub).not.toHaveBeenCalled()
    expect(mockFaTwitter).not.toHaveBeenCalled()
  })

  test("should render FaGithub", () => {
    mockUseThemeOption.mockImplementationOnce(() => {
      return {
        links: [
          {
            name: "GitHub",
            url: "/",
          },
        ],
      }
    })
    render(<ExternalLinks />)
    expect(mockFaLink).not.toHaveBeenCalled()
    expect(mockFaRss).not.toHaveBeenCalled()
    expect(mockFaGithub).toHaveBeenCalled()
    expect(mockFaTwitter).not.toHaveBeenCalled()
  })

  test("should render FaTwitter", () => {
    mockUseThemeOption.mockImplementationOnce(() => {
      return {
        links: [
          {
            name: "Twitter",
            url: "/",
          },
        ],
      }
    })
    render(<ExternalLinks />)
    expect(mockFaLink).not.toHaveBeenCalled()
    expect(mockFaRss).not.toHaveBeenCalled()
    expect(mockFaGithub).not.toHaveBeenCalled()
    expect(mockFaTwitter).toHaveBeenCalled()
  })
})
