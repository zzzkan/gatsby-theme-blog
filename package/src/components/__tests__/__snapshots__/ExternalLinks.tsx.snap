// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExternalLinks component > snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-0>*:not(style)~*:not(style) {
  margin-top: 0px;
  -webkit-margin-end: 0px;
  margin-inline-end: 0px;
  margin-bottom: 0px;
  -webkit-margin-start: 1px;
  margin-inline-start: 1px;
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: 2px solid transparent;
  outline-offset: 2px;
  padding: 0px;
  border-radius: full;
  font-size: xl;
}

<div
    class="chakra-stack emotion-0"
  >
    <a
      aria-label="Profile"
      class="chakra-button emotion-1"
      href="/profile"
      rel="noreferrer"
      target="_blank"
    />
    <a
      aria-label="RSS"
      class="chakra-button emotion-1"
      href="/rss"
      rel="noreferrer"
      target="_blank"
    />
    <a
      aria-label="GitHub"
      class="chakra-button emotion-1"
      href="/github"
      rel="noreferrer"
      target="_blank"
    />
    <a
      aria-label="Twitter"
      class="chakra-button emotion-1"
      href="/twitter"
      rel="noreferrer"
      target="_blank"
    />
  </div>
</DocumentFragment>
`;
